processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java

```diff
@@ -0,0 +1,249 @@
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation.first;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.primitives.Doubles;
+import com.google.common.primitives.Longs;
+import com.metamx.common.StringUtils;
+import io.druid.collections.SerializablePair;
+import io.druid.query.aggregation.Aggregator;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.AggregatorFactoryNotMergeableException;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.Column;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+public class DoubleFirstAggregatorFactory extends AggregatorFactory
+{
+  private static final byte CACHE_TYPE_ID = 16;
+
+  private final String fieldName;
+  private final String name;
+
+  @JsonCreator
+  public DoubleFirstAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName
+  )
+  {
+    Preconditions.checkNotNull(name, "Must have a valid, non-null aggregator name");
+    Preconditions.checkNotNull(fieldName, "Must have a valid, non-null fieldName");
+
+    this.name = name;
+    this.fieldName = fieldName;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    return new DoubleFirstAggregator(
+        name, metricFactory.makeFloatColumnSelector(fieldName),
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME)
+    );
+  }
+
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    return new DoubleFirstBufferAggregator(
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+        metricFactory.makeFloatColumnSelector(fieldName)
+    );
+  }
+
+  @Override
+  public Comparator getComparator()
+  {
+    return new Comparator()
+    {
+      @Override
+      public int compare(Object o1, Object o2)
+      {
+        return Doubles.compare(((SerializablePair<Long, Double>) o1).rhs, ((SerializablePair<Long, Double>) o2).rhs);
+      }
+    };
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    return (((SerializablePair<Long, Double>) lhs).lhs <= ((SerializablePair<Long, Double>) rhs).lhs) ? lhs : rhs;
```

Comment:
Combine uses a different compare function than the Comparator above, as combine compares time on the lhs while getComparator compares the value on the rhs. It's probably still a good idea to put the combine comparator as static object since (...lhs).lhs seems pretty confusing